dataflow/src/main/java/org/checkerframework/dataflow/cfg/DOTCFGVisualizer.java

```diff
@@ -38,442 +27,272 @@
-                this.sbDigraph.append(
-                        "Process order: "
-                                + processOrder.get(v).toString().replaceAll("[\\[\\]]", "")
-                                + "\\n");
+                sbDotNodes
+                        .append(getProcessOrderSimpleString(processOrder.get(v)))
+                        .append(leftJustifiedTerminator);
             }
-            visualizeBlock(v, analysis);
+            sbDotNodes.append(visualizeBlock(v, analysis));
         }
 
-        this.sbDigraph.append("\n");
+        sbDotNodes.append(lineSeparator);
+        return sbDotNodes.toString();
     }
 
-    /** @return the file name used for DOT output. */
-    protected String dotOutputFileName(UnderlyingAST ast) {
-        StringBuilder srcloc = new StringBuilder();
-
-        StringBuilder outfile = new StringBuilder(outdir);
-        outfile.append('/');
-        if (ast.getKind() == UnderlyingAST.Kind.ARBITRARY_CODE) {
-            CFGStatement cfgs = (CFGStatement) ast;
-            String clsname = cfgs.getClassTree().getSimpleName().toString();
-            outfile.append(clsname);
-            outfile.append("-initializer-");
-            outfile.append(ast.hashCode());
-
-            srcloc.append('<');
-            srcloc.append(clsname);
-            srcloc.append("::initializer::");
-            srcloc.append(((JCTree) cfgs.getCode()).pos);
-            srcloc.append('>');
-        } else if (ast.getKind() == UnderlyingAST.Kind.METHOD) {
-            CFGMethod cfgm = (CFGMethod) ast;
-            String clsname = cfgm.getClassTree().getSimpleName().toString();
-            String methname = cfgm.getMethod().getName().toString();
-            outfile.append(clsname);
-            outfile.append('-');
-            outfile.append(methname);
-
-            srcloc.append('<');
-            srcloc.append(clsname);
-            srcloc.append("::");
-            srcloc.append(methname);
-            srcloc.append('(');
-            srcloc.append(cfgm.getMethod().getParameters());
-            srcloc.append(")::");
-            srcloc.append(((JCTree) cfgm.getMethod()).pos);
-            srcloc.append('>');
-        } else {
-            throw new BugInCF("Unexpected AST kind: " + ast.getKind() + " value: " + ast);
-        }
-        outfile.append('-');
-        outfile.append(checkerName);
-        outfile.append(".dot");
-
-        // make path safe for Windows
-        String out = outfile.toString().replace("<", "_").replace(">", "");
+    @Override
+    protected String addEdge(long sId, long eId, String flowRule) {
+        return "    " + sId + " -> " + eId + " [label=\"" + flowRule + "\"];" + lineSeparator;
+    }
 
-        generated.put(srcloc.toString(), out);
+    @Override
+    public String visualizeBlock(Block bb, @Nullable Analysis<A, S, T> analysis) {
+        return super.visualizeBlockHelper(
+                bb,
+                analysis,
+                " \",];" + lineSeparator,
```

Comment:
It is very confusing that the `"` and `[` of `visualizeNodes` is closed here in a different method, without any documentation about the relationship between them or their contracts.  I suggest that you move that part of these strings into the `visualizeNodes` routine.